src/terms.jl

```diff
@@ -398,18 +376,52 @@ Base.show(io::IO, mime::MIME"text/plain", t::MatrixTerm; prefix="") =
 # operators on Terms that create new terms:
 
 
-Base.:~(lhs::TermOrTerms, rhs::TermOrTerms) = FormulaTerm(lhs, rhs)
+Base.:~(lhs::TermOrTerms, rhs::TermOrTerms) = FormulaTerm(lhs, cleanup(rhs))
 
-Base.:&(terms::AbstractTerm...) = InteractionTerm(terms)
 Base.:&(term::AbstractTerm) = term
-Base.:&(it::InteractionTerm, terms::AbstractTerm...) = InteractionTerm((it.terms..., terms...))
-
+Base.:&(a::AbstractTerm, b::AbstractTerm) = InteractionTerm((a,b))
+
+Base.:&(::ConstantTerm, b::AbstractTerm) = b
+Base.:&(a::AbstractTerm, ::ConstantTerm) = a
+
+# Avoid method ambiguities
+Base.:&(::ConstantTerm, b::InteractionTerm) = b
+Base.:&(a::InteractionTerm, ::ConstantTerm) = a
+Base.:&(a::ConstantTerm, ::ConstantTerm) = a
```

Comment:
`ConstantTerm` allows storing a constant value, which is completely ignored here. Is this intended?

Reply:
mostly laziness on my part; we could add a check that it stores a `1` and throw an error otherwise?

Reply:
Why not avoid storing anything in it then?

Reply:
Because it's what constant numeric values get lowered to during `@formula`, so we need to be able to access them for things like `@formula(poly(a, 3))`.

Reply:
OK. It would sound safer to check that the value is 1 then.

Reply:
The current behavior on 0.6.x is a bit inconsistent here between `@formula` and run-time:

```julia
julia> @formula(y ~ 1 & 2)
┌ Warning: Number 2 removed from interaction term (&)(2)
└ @ StatsModels ~/.julia/packages/StatsModels/fK0P3/src/formula.jl:158
FormulaTerm
Response:
  y(unknown)
Predictors:

julia> term(1) & term(2)
1 & 2
```

So whatever we do here, we're going to be inconsistent with ONE of the two paths lol.

As it currently stands in this PR, we're closer to what happens in `@formula` on main: constant terms are removed from interactions regardless of their value.  Should we be making it an error to have a constant that's not `1` in an interaction?  That seems a bit aggressive to me but I also have a hard time thinking of cases when that would actually come up, so making it an error might not be that bad of an idea since this is a breaking change anyway, and it makes it possible to change the behavior later with less risk that someone is relying on essentially undefined behavior...

Reply:
Yes better throw errors in absurd cases rather than risk hiding bugs. So with the new state of the PR an error is thrown in both cases above, right?

Reply:
yup, any time a ConstantTerm is involved in a `&` we check that it's `1` and throw an error otherwise!